### PR TITLE
build(nsis): Write registry keys to HKLM instead of HKCU, Install shortcuts for all users, Fix INSTALLDIR removal bug

### PR DESCRIPTION
**Pull request by sitiom**
This is the proper root key to write registries for machine-wide installation. Winget was having problems due to this inconsistency: https://github.com/microsoft/winget-pkgs/issues/47655

### CODE REVIEW

**Comment by jamescowens** [comment]
Thanks! :)

UT. Concept ACK. If you are going to change the key root, you will need to handle the removal of stale keys in the old location upon install. Your current PR will not do that.

This will also produce a requirement for user elevation for the install I think. Gridcoin itself is not to be run under elevated privileges.

**Comment by sitiom** [comment]
> This will also produce a requirement for user elevation for the install I think. Gridcoin itself is not to be run under elevated privileges.

The current installer **does** require user elevation (it installs to Program Files). See the UAC icon:
image

**Comment by jamescowens** [comment]
Agreed. Still need to deal with removing the old stale entries leftover from the old install on an overlay install using the new installer. You also should think about the downgrade case in case someone temporarily downgrades to make sure the extra keys in HKLM (from the perspective of the old install) don't cause a problem. There are user specific settings written to the registry regarding Window Size, etc that should be maintained in HKCU.

**Comment by sitiom** [comment]
I think the easiest solution would be to run the uninstaller first before installing the new version to remove old user entries and shortcuts.

**Comment by jamescowens** [comment]
You can't depend on people doing that. You would have to put a check to not allow the installer to run if the uninstaller had not run before. You would also have to ensure people that the data directory would not be affected. Many people can get confused between the data directory and the program executable directory.

**Comment by jamescowens** [comment]
BTW, Gridcoin has its own internal update check. Not sure why we need to get an external one to work.

**Comment by sitiom** [comment]
> You can't depend on people doing that. You would have to put a check to not allow the installer to run if the uninstaller had not run before.

This installer can also prompt to uninstall first if it detects the old HKCU entry; that would be more convenient.

> BTW, Gridcoin has its own internal update check. Not sure why we need to get an external one to work.

Because of the installation inconsistency, since the current installer writes to HKCU, other users will not be able to see the start menu shortcuts and installation entry in the Control Panel despite it being installed machine-wide. That's another problem that needs to be solved.

**Comment by sitiom** [comment]
And, oh yeah. I found a bug in the installer. I was confused why my installation directory disappeared when I hadn't done anything yet, so I found out that just running the installer from the start will delete the installation directory due to these lines of code:

https://github.com/gridcoin-community/Gridcoin-Research/blob/3f2a0955d1a371a2c6d1e2deb8d97760077551b6/share/setup.nsi.in#L178-L180

Also, the `uninst.exe` line doesn't do anything, as the uninstaller file is `uninstaller.exe`. I'll also attempt to fix this in this PR.

**Comment by sitiom** [comment]
> This installer can also prompt to uninstall first if it detects the old HKCU entry; that would be more convenient.

Done. Fixed the bug mentioned in https://github.com/gridcoin-community/Gridcoin-Research/pull/2643#issuecomment-1445678304 as well. If the HKCU keys are found, it will prompt the user to uninstall first:
image

Installation will immediately proceed afterwards after a successful uninstallation. Otherwise, it will abort. Silent installation using the `/S` switch also works seamlessly.

**Comment by jamescowens** [comment]
Very good. I will look at this closely today! :)

File: share/setup.nsi.in [hunk 1]
```diff
@@ -159,30 +159,41 @@ SectionEnd
...
-    rununin:
-      Exec $INSTDIR\uninst.exe
-      Delete $INSTDIR\*"
+    ; Check for HKCU Path registry and prompt for uninstallation if found
+    ReadRegStr $0 HKCU "${REGKEY}" "Path"
+    ${IfThen} $0 == "" ${|} Return ${|}
+    ${IfNot} ${Cmd} \
+        'MessageBox MB_YESNO|MB_ICONEXCLAMATION \
+          "An old version is found to be installed on your system. \
+          Uninstallation will proceed first to migrate registry entries and shortcuts. Continue?" \
```
Comment (jamescowens) [inline share/setup.nsi.in]: Can we be a little more terse? Maybe

An old version is found on your system and will be uninstalled first to migrate registry entries and shortcuts.  Your wallet data will not be affected. Continue?

**Review by jamescowens — CHANGES_REQUESTED** [review]
tACK except for the changes I asked for in the dialog box. Good work.

File: share/setup.nsi.in (hunk 1 above)
Comment (sitiom) [reply share/setup.nsi.in]: Done.

**Review by jamescowens — APPROVED** [review]
tACK.

**Pull request merged**